tests/validation/cattlevalidationtest/core/test_services.py

```diff
@@ -0,0 +1,914 @@
+            restart_policy["name"]
+        assert inspect["HostConfig"]["RestartPolicy"]["MaximumRetryCount"] == \
+            restart_policy["maximumRetryCount"]
+        assert inspect["Config"]["Cmd"] == command
+        assert inspect["Config"]["Memory"] == memory
+        assert "TEST_FILE=/etc/testpath.conf" in inspect["Config"]["Env"]
+        assert inspect["Config"]["CpuShares"] == cpu_shares
+
+    delete_all(client, [env])
+
+
+def test_services_port_and_link_options(super_client, client, managed_network,
+                                        socat_containers):
+
+    hosts = client.list_host(kind='docker', removed_null=True, state="active")
+
+    host = hosts[0]
+    link_host = hosts[0]
+
+    link_name = "WEB1"
+    link_port = 80
+    exposed_port = 9999
+
+    link_container = client.create_container(
+        imageUuid=LB_IMAGE_UUID,
+        environment={'CONTAINER_NAME': link_name},
+        networkIds=[managed_network.id],
+        name=random_str(),
+        requestedHostId=host.id
+        )
+
+    link_container = client.wait_success(link_container)
+
+    launch_config = {"imageUuid": SSH_IMAGE_UUID,
+                     "ports": [str(exposed_port)+":22/tcp"],
+                     "instanceLinks": {
+                         link_name:
+                             link_container.id},
+                     "requestedHostId": link_host.id,
+                     "networkIds": [managed_network.id]
+                     }
+
+    service, env = create_env_and_svc(client, launch_config, 1)
+
+    env = env.activateservices()
+    service = client.wait_success(service, 300)
+
+    container_name = env.name + "_" + service.name + "_1"
+    containers = client.list_container(name=container_name, state="running")
+    assert len(containers) == 1
+    con = containers[0]
+
+    time.sleep(10)
+    validate_exposed_port_and_container_link(super_client, con, link_name,
+                                             link_port, exposed_port)
+
+    delete_all(client, [env, link_container])
+
+
+def test_environment_activate_decativate_delete(super_client,
+                                                client, managed_network,
+                                                socat_containers):
+
+    launch_config = {"imageUuid": TEST_IMAGE_UUID}
+
+    scale = 1
+
+    service1, env = create_env_and_svc(client, launch_config,
+                                       scale)
+
+    service2 = create_svc(client, env, launch_config, scale)
+
+    # Environment Activate Services
+    env = env.activateservices()
+
+    service1 = client.wait_success(service1, 300)
+    assert service1.state == "active"
+    check_container_in_service(super_client, service1)
+
+    service2 = client.wait_success(service2, 300)
+    assert service2.state == "active"
+    check_container_in_service(super_client, service2)
+
+    # Environment Deactivate Services
+    env = env.deactivateservices()
+
+    time.sleep(60)
```

Comment:
Removed time.sleep and instead added a method - wait_until_instances_get_stopped() to wait until all the instances of the service gets to stopped state